### PR TITLE
Add conformer RMSD result filter

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description

This PR adds a new filter which aims to retain up to a maximum number of maximally distinct conformers for each molecule in an optimization or a basic result collection while conformers similar to within an RMSD threshold are discarded.

## Status
- [x] Ready to go

### CODE REVIEW

File: openff/qcsubmit/results/filters.py [hunk 1]
```diff
@@ -158,19 +160,20 @@ def _apply(self, result_collection: "T") -> "T":
...
+          Torsion drive datasets / entries will be skipped.
     """
 
     @abc.abstractmethod
     def _filter_function(
         self, entries: List[Tuple["_BaseResult", RecordBase, Molecule, str]]
-    ) -> Tuple["_BaseResult", str]:
-        """A method which should reduce a set of results down to a single entry based on some property of the QC
-        calculation.
+    ) -> List[Tuple["_BaseResult", str]]:
```
Comment (jthorton) [inline openff/qcsubmit/results/filters.py:174]: Ahh good catch!

File: openff/qcsubmit/results/filters.py [hunk 2]
```diff
@@ -194,13 +197,16 @@ def _apply(self, result_collection: "T") -> "T":
         filtered_results = defaultdict(list)
 
         for entries in entries_by_inchikey.values():
-            result, address = self._filter_function(
+
+            results_and_addresses = self._filter_function(
```
Comment (jthorton) [inline openff/qcsubmit/results/filters.py:201]: Nice!

**Review by jthorton — COMMENTED** [review]
Thanks for adding this I think this will be very useful. One thing I am not sure about is, what happens if a molecule is in two different datasets in a different atom ordering when trying to compare the RMSD? Should the molecule given by the `to_records` function put all of the molecules into canonical order depending on what toolkit is being used to do the filtering?

**Comment by SimonBoothroyd** [comment]
Thanks @jthorton for the review!

> Should the molecule given by the to_records function put all of the molecules into canonical order depending on what toolkit is being used to do the filtering?

Great catch! I've fixed this in fd61bf9

**Pull request merged**